File: chat/src/main/java/glitch/chat/object/irc/Prefix.java
package glitch.chat.object.irc;

import lombok.AccessLevel;
import lombok.Data;
import lombok.RequiredArgsConstructor;

import javax.annotation.Nullable;

/**
 * @author Damian Staszewski
 * @since 0.1.0
 */
@Data
@RequiredArgsConstructor(access = AccessLevel.PRIVATE)
public class Prefix {
    private final String raw;
    @Nullable
    private final String nick;
    @Nullable
    private final String user;
    private final String host;

    /**
     * @return
     */
    public static Prefix empty() {
        return fromRaw(":tmi.twitch.tv");
    }

    /**
     * @param rawPrefix
     * @return
     */
    public static Prefix fromRaw(String rawPrefix) {
        if (rawPrefix.matches(":(?:.*)tmi.twitch.tv")) {
            throw new IllegalArgumentException("The RAW Prefix is invalid! PREFIX: " + rawPrefix);
        }
        String prefix = rawPrefix.substring(1);
        String nick = null;
        String user = null;
        String host;


        if (prefix.contains("@")) {
            String[] nh = rawPrefix.split("@");
            host = nh[1];
            if (nh[0].contains("!")) {
                String[] nu = nh[0].split("!");
                nick = nu[0];
                user = nu[1];
            } else {
                nick = nh[0];
            }
        } else {
            host = prefix;
        }

        return new Prefix(rawPrefix, nick, user, host);
    }
}

File: chat/src/main/kotlin/glitch/chat/irc/Prefix.kt
package glitch.chat.irc

/**
 *
 * @author Damian Staszewski [damian@stachuofficial.tv]
 * @version %I%, %G%
 * @since 0.1.0
 */
data class Prefix(val raw: String, val nick: String? = null, val user: String? = null, val host: String) {
    override fun toString(): String = raw

    companion object {
        /**
         * @param rawPrefix
         * @return
         */
        fun fromRaw(rawPrefix: String): Prefix {
            if (!rawPrefix.matches(":(?:.*)tmi.twitch.tv".toRegex())) {
                throw IllegalArgumentException("The RAW Prefix is invalid! PREFIX: $rawPrefix")
            }
            val prefix = rawPrefix.substring(1)
            var nick: String? = null
            var user: String? = null
            val host: String


            if (prefix.contains("@")) {
                val nh = rawPrefix.split("@".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
                host = nh[1]
                if (nh[0].contains("!")) {
                    val nu = nh[0].split("!".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
                    nick = nu[0]
                    user = nu[1]
                } else {
                    nick = nh[0]
                }
            } else {
                host = prefix
            }

            return Prefix(rawPrefix, nick, user, host)
        }

        fun empty() = Prefix(":tmi.twitch.tv", host = "tmi.twitch.tv")
    }
}